PSYKHANOOL/PSYKHANOOL.h

```diff
@@ -0,0 +1,664 @@
+/*******************************************************************************
+The MIT License (MIT)
+
+Copyright (c) 2015 Dmitry "Dima" Korolev <dmitry.korolev@gmail.com>
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+// Current-friendly data manipulation types:
+//
+// * bool Exists(x) : non-throwing, always `true` for non-optionals.
+// * T Value(x)     : can throw `NoValue` if and only if `Exists(x)` is `false`.
+//
+// Current-friendly container types.
+//
+// * Vector<T> <=> std::vector<T>
+//   Empty(), Size(), operator[](i), PushBack(x), PopBack() [, iteration].
+//
+// * OrderedDictionary<T> <=> std::map<T_KEY, T>
```

Comment:
IMHO `UnorderedDictionary` is more frequent usecase (and `unordered_map` has much better performance).